src/main/java/org/kiwiproject/registry/eureka/client/EurekaParser.java

```diff
@@ -0,0 +1,84 @@
+
+import static com.google.common.collect.Lists.newArrayList;
+import static java.util.Objects.nonNull;
+import static java.util.stream.Collectors.toList;
+import static org.apache.commons.collections4.MapUtils.getString;
+
+import lombok.experimental.UtilityClass;
+import org.kiwiproject.registry.eureka.common.EurekaInstance;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+@UtilityClass
+public class EurekaParser {
+
+    @SuppressWarnings("unchecked")
+    public static List<EurekaInstance> parseEurekaResponse(Map<String, Object> response) {
+        var applications = ((Map<String, Object>) response.get("applications")).get("application");
+
+        var eurekaInstances = new ArrayList<EurekaInstance>();
+
+        if (nonNull(applications)) {
+            if (applications instanceof List<?>) {
+                for (Map<String, Object> application : (List<Map<String, Object>>) applications) {
+                    var instances = application.get("instance");
+                    eurekaInstances.addAll(parseInstances(instances));
+                }
+            } else {
+                eurekaInstances.addAll(parseInstances(((Map<String, Object>) applications).get("instance")));
+            }
+        }
+
+        return eurekaInstances;
+    }
+
+    @SuppressWarnings("unchecked")
+    private static List<EurekaInstance> parseInstances(Object instanceOrInstanceList) {
+        if (instanceOrInstanceList instanceof List<?>) {
+            var instanceList = (List<Map<String, Object>>) instanceOrInstanceList;
+            return instanceList.stream()
+                    .map(EurekaParser::buildInstance)
+                    .collect(toList());
+        }
+
+        var instance = (Map<String, Object>) instanceOrInstanceList;
+        return newArrayList(buildInstance(instance));
+    }
+
+    @SuppressWarnings("unchecked")
+    public static EurekaInstance buildInstance(Map<String, Object> instanceData) {
+
+        var portMap = (Map<String, Object>) instanceData.get("port");
+        var securePortMap = (Map<String, Object>) instanceData.get("securePort");
+        var metadataMap = (Map<String, String>) instanceData.get("metadata");
+        var leaseInfoMap = (Map<String, Object>) instanceData.get("leaseInfo");
+
+        var leaseInfo = Map.of(
+                "renewalIntervalInSecs", leaseInfoMap.getOrDefault("renewalIntervalInSecs", 0),
+                "durationInSecs", leaseInfoMap.getOrDefault("durationInSecs", 0),
+                "registrationTimestamp", leaseInfoMap.getOrDefault("registrationTimestamp", 0L),
+                "lastRenewalTimestamp", leaseInfoMap.getOrDefault("lastRenewalTimestamp", 0L),
+                "evictionTimestamp", leaseInfoMap.getOrDefault("evictionTimestamp", 0L),
+                "serviceUpTimestamp", leaseInfoMap.getOrDefault("serviceUpTimestamp", 0L));
+
+        return EurekaInstance.builder()
+                .vipAddress(getString(instanceData, "vipAddress"))
```

Comment:
For example here is the implementation in `MapUtils`:

```java
    public static <K> String getString(final Map<? super K, ?> map, final K key) {
        if (map != null) {
            final Object answer = map.get(key);
            if (answer != null) {
                return answer.toString();
            }
        }
        return null;
    }
```

We clearly don't need the null check on the map (since we would have blown up way before getting here). So we could choose to do this:

```java
    private static <K> String getString(final Map<? super K, ?> map, final K key) {
        var value = map.get(key);
        return nonNull(value) ? value.toString() : null;
    }
```

which retains the behavior of `MapUtils` in terms of calling `toString()` to perform the String conversion.

OTOH, I guess I never really read the docs for `MapUtils.getString` but I'm not really sure we want to simply call `toString()` like that. The values we are trying to get from the map _should already be Strings_ (or null) and so we should be able to simply cast to String. In that case, we could simply do:

```java
    private static <K> String getString(final Map<? super K, ?> map, final K key) {
        var value = map.getOrDefault(key, null);
        return (String) value;
    }
```

Thoughts?

Reply:
Done